File: backend/app/services/customers.py
from app.database.connection import execute_query, execute_write
from fastapi import HTTPException

def get_customers_by_company(
    company_id: int,
    queue_status: str | None = None,
) -> list[dict]:
    query = """
        SELECT
            id,
            company_id,
            name,
            phone,
            queue_status,
            last_message,
            review_reason,
            created_at,
            updated_at
        FROM customers
        WHERE company_id = %(company_id)s
    """

    parameters = {
        "company_id": company_id,
    }

    if queue_status is not None:
        query += """
            AND queue_status = %(queue_status)s
        """
        parameters["queue_status"] = queue_status

    query += """
        ORDER BY updated_at DESC
    """

    return execute_query(query, parameters)

def mark_completed(customer_id: int) -> dict:
    query = """
        UPDATE customers
        SET
            queue_status = 'completed',
            review_reason = NULL,
            updated_at = NOW()
        WHERE id = %(customer_id)s
        RETURNING
            id,
            company_id,
            name,
            phone,
            queue_status,
            last_message,
            review_reason,
            created_at,
            updated_at;
    """

    result = execute_write(
        query,
        {
            "customer_id": customer_id,
        },
    )

    if result is None:
        raise HTTPException(
            status_code=404,
            detail="Customer not found.",
        )

    return result

File: frontend/src/api/customers.ts
import { apiRequest } from "./client";

export type QueueStatus = "review" | "completed";

interface CustomerApiResponse {
  id: number;
  company_id: number;
  name: string;
  phone: string | null;
  queue_status: QueueStatus;
  last_message: string | null;
  review_reason: string | null;
  created_at: string;
  updated_at: string;
}

export interface Customer {
  id: number;
  companyId: number;
  name: string;
  phone: string | null;
  queueStatus: QueueStatus;
  lastMessage: string | null;
  reviewReason: string | null;
  createdAt: string;
  updatedAt: string;
}

export async function getCustomers(
  companyId: number,
  queueStatus?: QueueStatus
): Promise<Customer[]> {
  const query = queueStatus
    ? `?queue_status=${encodeURIComponent(queueStatus)}`
    : "";

  const customers = await apiRequest<CustomerApiResponse[]>(
    `/companies/${companyId}/customers${query}`
  );

  return customers.map((customer) => ({
    id: customer.id,
    companyId: customer.company_id,
    name: customer.name,
    phone: customer.phone,
    queueStatus: customer.queue_status,
    lastMessage: customer.last_message,
    reviewReason: customer.review_reason,
    createdAt: customer.created_at,
    updatedAt: customer.updated_at,
  }));
}

export async function markCustomerCompleted(
  customerId: number
): Promise<Customer> {
  const customer = await apiRequest<CustomerApiResponse>(
    `/companies/customers/${customerId}/complete`,
    {
      method: "PATCH",
    }
  );

  return {
    id: customer.id,
    companyId: customer.company_id,
    name: customer.name,
    phone: customer.phone,
    queueStatus: customer.queue_status,
    lastMessage: customer.last_message,
    reviewReason: customer.review_reason,
    createdAt: customer.created_at,
    updatedAt: customer.updated_at,
  };
}